proto/substrait/algebra.proto

```diff
@@ -561,14 +561,63 @@ message WriteRel {
   }
 }
 
+// Equality joins (equi-joins) are a specialization of the general join where the join expression
+// is an series of equality comparisons between fields that are ANDed together.  The exact definition
+// of "equality" is flexible.
+message EquiJoinKey {
+  // The key to compare from the left table
+  Expression.FieldReference left = 1;
+  // The key to compare from the right table
+  Expression.FieldReference right = 2;
+  // Describes how to compare the two keys, defaults to the equal function
+  EqualityType equality = 3;
+
+  // Most equi-joins will use one of the following equality behaviors.  To avoid the complexity
+  // of a function lookup we define the common behaviors here.
+  enum SimpleEqualityType {
+    SIMPLE_EQUALITY_TYPE_UNSPECIFIED = 0;
+    // Returns true only if both values are equal and not null
+    SIMPLE_EQUALITY_TYPE_EQ = 1;
+    // Returns true if both values are equal and not null
+    // Returns true if both values are null
+    // Returns false if one value is null and the other value is not null
+    //
+    // This can be expressed as a = b OR (isnull(a) AND isnull(b))
+    SIMPLE_EQUALITY_TYPE_IS_NOT_DISTINCT_FROM = 2;
+    // Returns true if both values are equal and not null
+    // Returns true if either value is null
+    //
+    // This can be expressed as a = b OR isnull(a = b)
+    SIMPLE_EQUALITY_TYPE_MIGHT_EQUAL = 3;
+  }
+
+  // Describes how the relation should consider if two rows are a match
```

Comment:
Half the comments in this PR end in a period, the others don't.  Please make them consistent.